DataDog/datadog-agent
# [system-probe][NET-381] scope dns stats by domain

### What does this PR do?

This adds the option of scoping DNS stats by domain.

The `dnsStats` map earlier was one layer : `dnsKey` -> `dnsStats`
This PR introduces an additional layer : `dnsKey` -> `domain` -> `dnsStats`

So far, no visible difference has been noticed in CPU usage (the yellow one is the with domains):

<img width="576" alt="Screen Shot 2020-11-06 at 9 36 48 AM" src="https://user-images.githubusercontent.com/6791023/98384913-fdd51600-2013-11eb-8f07-f9627d677646.png">



### Gating

The feature is gated by `system_probe_config.collect_dns_domains` and the environment variable `DD_COLLECT_DNS_DOMAINS`.

Initially, I was planning to keep the codepath and data structures for **domain-scoped** stats completely separate from the existing **domain-agnostic** one.

However, that led to too much complexity
-  We could have two versions of `DNSStatKeeper` and use one based on config variable.
-  But then we would have to have duplicate data structures in `State` to support both types of data.

So this PR takes a different approach. It makes all the data structures and functions  **domain-scoped**, even when the config flag is turned off.  This is what happens differently based on whether the flag is turned off or not:
-  `dnsParser` leaves the domain field `nil`, so **all** stats get scoped by the same domain
```go
if p.collectDNSDomains {
    pktInfo.question = string(question.Name)
}
```
- `networkState`   fills up  (`conn.DNSSuccessfulResponses`, `conn.DNSTimeouts`) OR `conn.DNSStatsByDomain[domain]`

```go
for domain, dnsStats := range dnsStatsByDomain {
    if ns.collectDNSDomains {
        var ds DNSStats
	ds.DNSTimeouts = dnsStats.timeouts
        ...... 
	conn.DNSStatsByDomain[domain] = ds
    } else {
	conn.DNSSuccessfulResponses += dnsStats.countByRcode[DNSResponseCodeNoError]
	conn.DNSTimeouts += dnsStats.timeouts
        ......
   }
```

### Encoding

It is likely that the set of unique domains queried would be really small. But if store them separately for each connection, it would be lots of duplication.
Initially, we looked at reusing `encodedDNS` field which stores the `IP->domain` mapping.
But that approach was abandoned because of the complexity:
- The set of domains would not be same. For example, if a query for a domain resulted in `NXDomain`, it will show up in `DNSStats` but not in the `IP->domain` mapping. So injecting those extra domains would make the code complex.
- The pointers would have to be updated in each connection after the generation of  `encodedDND`

Instead, we simply have a new field (a list of `string`s) that contains all the domains and `DNSStats` stores pointers to positions in that list:
```
    "dnsStatsByDomain": {
        "0": {
          "dnsTimeouts": 0,
          "dnsSuccessLatencySum": "18505",
          "dnsFailureLatencySum": "0",
          "dnsCountByRcode": {
            "0": 1
          }
        }
      }
    "dnsStatsByDomain": {
        "1": {
          "dnsTimeouts": 0,
          "dnsSuccessLatencySum": "20261",
          "dnsFailureLatencySum": "0",
          "dnsCountByRcode": {
            "0": 1
          }
        }
  "domains": [
    "cnn.com",
    "bbc.com"
  ]
```


#### Next steps for encoding

To reuse `encodedDNS` we plan to take the following approach:

1. Go through all `connections` and for each `connection`, go through all `DNSStats` and create a set of `domains` (`pkg/network/encoding/format.go`, `pkg/network/encoding/protobuf.go`, `pkg/network/encoding/json.go`)

2. Change `dnsEncoder.encode` (https://github.com/DataDog/datadog-agent/blob/master/pkg/process/checks/net.go#L209) and pass that list of strings

3. From `encodedDNS`,  create a map  from domain->position.

4. Go through all `connections` and for each `connection`, go through all `DNSStats` and update the value for domain with the position in the encoded value.

### Describe your test plan

1. Make sure collections of `domains` is enabled:
```
system_probe_config:
  enabled: true
  collect_dns_domains: true
```
2. Run `system-probe`. `Dig` a few domains.
- `dig apple.com`
- `dig datadoghq.com`

3. Run the following command `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | python parser.py` where `parser.py` has the following contents: 


```python
import json
import sys

d = json.load(sys.stdin)
print "----------------------------------------------------------------------------"
for c in d['conns']:
    if not c['dnsStatsByDomain']:
        continue
    s =  "{: <5} {} {: >12}:{: <5} ->{: >16}:{: <5}"
    print s.format(
        c['pid'],
        c['type'],
        c['laddr']['ip'],
        c['laddr']['port'],
        c['raddr']['ip'],
        c['raddr']['port'],
    )
    for k, v in c['dnsStatsByDomain'].iteritems():
        s =  "{: >20} {: >2} {: >2} {: >2} {: >2} {: >2}"
        ec = v['dnsCountByRcode']
        success = 0
        failure = 0
        for codestr, count in ec.iteritems():
            code = int(codestr)
            if code == 0:
                success += 1
            else:
                failure += 1

        print s.format(
           d['domains'][int(k)],
           success,
           failure,
           v['dnsTimeouts'],
           v['dnsSuccessLatencySum'],
           v['dnsFailureLatencySum'],
        )
    print "----------------------------------------------------------------------------"
```
The output should be something like the following:

```
----------------------------------------------------------------------------
8199  udp    10.0.2.15:55238 ->        10.0.2.3:53
           apple.com  1  0  0 34892  0
----------------------------------------------------------------------------
8199  udp    10.0.2.15:36959 ->        10.0.2.3:53
       datadoghq.com  1  0  0 39834  0
----------------------------------------------------------------------------
```
